specfile/formatter.py

```diff
@@ -0,0 +1,132 @@
+
+    Only supports a small subset of Python AST that should be sufficient for use in __repr__().
+
+    Args:
+        expression: Python expression to reformat.
+        line_length_threshold: Threshold for line lengths. It's not a hard limit,
+          it can be exceeded in some cases.
+
+    Returns:
+        Formatted expression.
+
+    Raises:
+        SyntaxError if the expression is not parseable.
+        SpecfileException if there is an unsupported AST node in the expression.
+    """
+
+    def fmt(node, indent=0, prefix="", multiline=False):
+        result = " " * indent + prefix
+        if isinstance(node, ast.Constant):
+            result += repr(node.value)
+        elif isinstance(node, (ast.List, ast.Dict, ast.Call)):
+            if isinstance(node, ast.List):
+                start, end = "[", "]"
+                items = [(None, e) for e in node.elts]
+                delimiter = None
+            elif isinstance(node, ast.Dict):
+                start, end = "{", "}"
+                items = [(fmt(k, 0), v) for k, v in zip(node.keys, node.values)]
+                delimiter = ": "
+            elif isinstance(node, ast.Call):
+                start, end = f"{node.func.id}(", ")"
+                items = [(None, a) for a in node.args] + [
+                    (kw.arg, kw.value) for kw in node.keywords
+                ]
+                delimiter = "="
+            result += start
+            if multiline:
+                result += "\n"
+            while items:
+                key, value = items.pop(0)
+                result += fmt(
+                    value,
+                    indent + 4 if multiline else 0,
+                    f"{key}{delimiter}" if key else "",
+                )
+                if multiline:
+                    result += ",\n"
+                elif items:
+                    result += ", "
+            if multiline:
+                result += " " * indent
+            result += end
+        else:
+            raise SpecfileException(
+                f"Unsupported AST node: ast.{node.__class__.__name__}"
+            )
+        if not multiline and len(result) > line_length_threshold:
+            result = fmt(node, indent, prefix, True)
+        return result
+
+    def find_matching_bracket(value, index):
+        level = 0
+        for i in range(index, len(value)):
+            if value[i] == ">":
+                level -= 1
+                if level <= 0:
+                    return i + 1
+            elif value[i] == "<":
+                level += 1
+        return None
+
+    placeholders = []
+    while True:
+        try:
+            tree = ast.parse(expression, mode="eval")
+        except SyntaxError as e:
+            # some objects (e.g. enums) are not represented by a valid expression,
+            # replace their representation with placeholders
+            # (assume such a representation is enclosed in <> and doesn't spawn
+            # across multiple lines)
```

Comment:
Actually, this assumption is rather naive, as the representation clearly can be multiline:

```python
>>> spec.prep
<bound method Specfile.prep of Specfile(
    PosixPath('/home/nforro/tmp/rpm-specs/python-specfile.spec'),
    PosixPath('/home/nforro/tmp/rpm-specs'),
    False,
    [],
    False,
)>
```

But it shouldn't matter, because nothing like this should appear in repr of any of specfile classes. We can extend this in the future to make it more universal.